src/Microsoft.ServiceFabric.Actors/Runtime/MigrationWorker.cs

```diff
@@ -0,0 +1,232 @@
+// ------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.ServiceFabric.Actors.Runtime
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Fabric;
+    using System.IO;
+    using System.Net.Http;
+    using System.Runtime.Serialization;
+    using System.Text;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using System.Xml;
+    using Microsoft.ServiceFabric.Actors.Generator;
+    using Microsoft.ServiceFabric.Actors.Migration.Models;
+    using Microsoft.ServiceFabric.Data.Collections;
+
+    internal class MigrationWorker
+    {
+        private static readonly HttpClient Client = new HttpClient();
+        private KVStoRCMigrationActorStateProvider stateProvider;
+        private IReliableDictionary2<string, byte[]> metadataDict;
+        private StatefulServiceInitializationParameters initParams;
+        private long chunkSize;
+        private long itemsPerEnumeration;
+        private string endpoint;
+
+        public MigrationWorker(KVStoRCMigrationActorStateProvider stateProvider, ActorTypeInformation actorTypeInfo)
+        {
+            this.stateProvider = stateProvider;
+            this.initParams = this.stateProvider.GetInitParams();
+            this.GetUserSettingsOrDefault(actorTypeInfo);
+        }
+
+        internal async Task StartMigrationWorker(MigrationPhase phase, long startSN, long endSN, int workerIdentifier, CancellationToken cancellationToken)
+        {
+            this.metadataDict = await this.stateProvider.GetMetadataDictionaryAsync();
+            if (endSN - startSN < this.itemsPerEnumeration || this.itemsPerEnumeration == -1)
+            {
+                var dataToMigrate = await this.GetEnumeratedDataFromKvsAsync(phase, startSN, endSN - startSN);
+                await this.stateProvider.SaveStatetoRCAsync(dataToMigrate, cancellationToken);
+                await this.UpdateMetadataOnCompletion(phase, workerIdentifier, endSN);
+            }
+            else
+            {
+                var lastUpdatedSN = startSN - 1;
+                var endSNForChunk = lastUpdatedSN + this.itemsPerEnumeration;
+                while (endSNForChunk <= endSN)
+                {
+                    var dataToMigrate = await this.GetEnumeratedDataFromKvsAsync(phase, lastUpdatedSN + 1, this.itemsPerEnumeration);
+                    await this.stateProvider.SaveStatetoRCAsync(dataToMigrate, cancellationToken);
```

Comment:
LastAppliedSN should be updated in the same txn that is used to hydrate the actual actor data

Reply:
Updated metadata update for lastAppliedSN to happen during SaveStateAsync